Sources/StreamChat/WebSocketClient/Engine/URLSessionWebSocketEngine.swift

```diff
@@ -5,40 +5,54 @@
 import Foundation
 
 @available(iOS 13, *)
-class URLSessionWebSocketEngine: NSObject, WebSocketEngine, URLSessionDataDelegate, URLSessionWebSocketDelegate {
-    private var task: URLSessionWebSocketTask?
+class URLSessionWebSocketEngine: NSObject, WebSocketEngine {
+    private weak var task: URLSessionWebSocketTask? {
+        didSet {
+            oldValue?.cancel()
+        }
+    }
     
     let request: URLRequest
-    private var session: URLSession!
+    private var session: URLSession?
     let delegateOperationQueue: OperationQueue
+    let sessionConfiguration: URLSessionConfiguration
+    var urlSessionDelegateHandler: URLSessionDelegateHandler?
     
     var callbackQueue: DispatchQueue { delegateOperationQueue.underlyingQueue! }
     
     weak var delegate: WebSocketEngineDelegate?
     
     required init(request: URLRequest, sessionConfiguration: URLSessionConfiguration, callbackQueue: DispatchQueue) {
         self.request = request
+        self.sessionConfiguration = sessionConfiguration
         
         delegateOperationQueue = OperationQueue()
         delegateOperationQueue.underlyingQueue = callbackQueue
-        
+
         super.init()
-        
+    }
+    
+    func connect() {
+        urlSessionDelegateHandler = makeURLSessionDelegateHandler()
+
         session = URLSession(
             configuration: sessionConfiguration,
-            delegate: self,
+            delegate: urlSessionDelegateHandler,
             delegateQueue: delegateOperationQueue
         )
-    }
-    
-    func connect() {
-        task = session.webSocketTask(with: request)
+
+        task = session?.webSocketTask(with: request)
         doRead()
         task?.resume()
     }
     
     func disconnect() {
         task?.cancel(with: .normalClosure, reason: nil)
+        session?.invalidateAndCancel()
+
+        session = nil
+        task = nil
+        urlSessionDelegateHandler = nil
```

Comment:
No need to nil these values given the current code

Reply:
Unfortunately, it is. There are some leaks inside the Foundation library when using the WebSockets API, so this makes sure that at least our objects are freed from memory.

Reply:
Oh true, forgot how amazingly URLSession was built 😂